### ISSUE TITLE
Another approach to theming?

### ISSUE THREAD

**Issue by bartjkdp** [Project-Meerkat/concierge#51]
@lucbelliveau and I created a new way of adding themes through pip packages, but I can imagine this is a bit of hassle to setup for newcomers and also requires a custom CI pipeline for both GC and Pleio. 

We could also use a "mono-repo" approach where we place the pleio and gc themes in separate folders in `themes/` of this repository. This makes it a bit easier for newcomers to contribute and it is also easier to keep updates of concierge "in sync" with the themes.

I would love to have your opionions @lucbelliveau @markwooff @bryan-robitaille

**Comment by lucbelliveau** [comment]
I'm not sure moving the theme apps into the repo solves the problems you mention... You still need to configure which theme to use in CI... and for the deploy...

**Comment by lucbelliveau** [comment]
What use cases are the configurable themes intended to serve?  Django applications can override core functionality so you would still be able to create this type of theme, as a separate theme package...  but I'm not sure creating a CMS is a worthwhile endeavor for this project... if the intent is to have end-users modify the themes, I would recommend exploring options like `django-cms`.... this could easily be integrated via a theme package as well...   at the end of the day, I think all organizations who will want to use this project will have their own custom requirements that they will want to integrate... having those in a separate package will help keep the core clean and re-usable... at least that was my thinking when packages were suggested....  maybe the word "theme" makes it sound just to superficial... maybe the packages could be referred to as "implementations" or something like that.

**Comment by bryan-robitaille** [comment]
Use case was along the lines of:
Company X needs a light auth service that supports some of the latest technologies.  Doesn't want to spend development time on accessibility and UX and wants a solution that is already vetted in those areas.  Doesn't need to meet full company CLF but wants to have logo and color pallet match existing web env.

**Comment by lucbelliveau** [comment]
I would say the functionality for that use case is a good fit for the generic theme...  a quick way to get up and running.
Both pleio and GC already have customizations, or are likely to get some, that extend core beyond what would be possible with a generic theme anyway... (at least not without significant effort for little reward, in my opinion)


null

**Comment by bryan-robitaille** [comment]
How should we handle functionality differences?  For example a new functionality is added tot he core which is displayed through the views and templates.  The template packages would need to be constantly checking the change log of the core code to adapt their 'themes' no?

**Comment by lucbelliveau** [comment]
Working on the same code base with divergent requirements will lead to more problems I think... like repos that will be very difficult to sync....  at least this way changes to themes wont break downstream... 
Changes to core will certainly require a certain level of discipline... but that's not a bad thing IMO...  not with the goals of the project...  better release strategies, documentation... etc.  I think with this strategy you could do daily pulls from upstream... with not a lot of work.


null

**Comment by bartjkdp** [comment]
@bryan-robitaille Thanks for your comments! I think one of the reasons why we choose the themes approach is because both HTML and CSS of the GC theme evolved really differently from the Pleio implementation. The GC theme uses Semantic UI as basis, the Pleio theme uses Materialize CSS. This difference we probably cannot bridge with admin configuration.

I think it would be great if we manage to settle on 1 HTML/CSS implementation and configurable parameters through the admin. We can try to evolve the generic theme to see if we can manage to get there. Otherwise we are probably stuck with having our own themes.

Still think it would be a great feature to allow for custom themes as well. Starting a custom theme also infers you have to keep it up-to-date with upstream. That is the pain of having your own theme ;)

**Comment by bryan-robitaille** [comment]
Good point @bartjkdp .  I'd like to explore the possibility of being able to use a single implementation of HTML/CSS.  Would it be possible to provide a delta of what would be required to modify the generic theme in order to make it compatible with the required Pleio branding/templates?
I don't think the UI with change heavily in the coming years but I do see the potential for more functionality.  For example the inclusion of an introspection endpoint, multiple levels of identity assurance and associated registration, ability to pull user information from an external profile (resource server), ability to link accounts to other identity providers, etc.
